From: Kefeng Wang <wangkefeng.wang@huawei.com>
Subject: mm: introduce vma_is_stack() and vma_is_heap()
Date: Wed, 12 Jul 2023 22:38:27 +0800

Patch series "mm: convert to vma_is_heap/stack()".

Add vma_is_stack() and vma_is_heap() helper and use them to simplify code.


This patch (of 5):

Introduce the two helpers for general use.

Link: https://lkml.kernel.org/r/20230712143831.120701-1-wangkefeng.wang@huawei.com
Link: https://lkml.kernel.org/r/20230712143831.120701-2-wangkefeng.wang@huawei.com
Signed-off-by: Kefeng Wang <wangkefeng.wang@huawei.com>
Cc: Alex Deucher <alexander.deucher@amd.com>
Cc: Arnaldo Carvalho de Melo <acme@kernel.org>
Cc: "Christian König" <christian.koenig@amd.com>
Cc: Daniel Vetter <daniel@ffwll.ch>
Cc: David Airlie <airlied@gmail.com>
Cc: Eric Paris <eparis@parisplace.org>
Cc: Felix Kuehling <Felix.Kuehling@amd.com>
Cc: "Pan, Xinhui" <Xinhui.Pan@amd.com>
Cc: Paul Moore <paul@paul-moore.com>
Cc: Peter Zijlstra <peterz@infradead.org>
Cc: Stephen Smalley <stephen.smalley.work@gmail.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 include/linux/mm.h |   12 ++++++++++++
 1 file changed, 12 insertions(+)

--- a/include/linux/mm.h~mm-introduce-vma_is_stack-and-vma_is_heap
+++ a/include/linux/mm.h
@@ -838,6 +838,18 @@ static inline bool vma_is_anonymous(stru
 	return !vma->vm_ops;
 }
 
+static inline bool vma_is_heap(struct vm_area_struct *vma)
+{
+       return vma->vm_start <= vma->vm_mm->brk &&
+		vma->vm_end >= vma->vm_mm->start_brk;
+}
+
+static inline bool vma_is_stack(struct vm_area_struct *vma)
+{
+       return vma->vm_start <= vma->vm_mm->start_stack &&
+	       vma->vm_end >= vma->vm_mm->start_stack;
+}
+
 static inline bool vma_is_temporary_stack(struct vm_area_struct *vma)
 {
 	int maybe_stack = vma->vm_flags & (VM_GROWSDOWN | VM_GROWSUP);
_
